[Clangd] Add AST search capabilities from clang-query

Febbe opened the pull request:
This is the first stage implementation of the https://github.com/clangd/clangd/issues/1945 proposal.

A search can be issued via the "textDocument/searchAST" JsonRPC function. 
It currently requires the search query and the source code url.

A query has to be in the form of the [Clang Matcher eDSL](https://clang.llvm.org/docs/LibASTMatchers.html).
E.g.: `declRefExpr(to(namedDecl(hasName("bob"))))` to match
`bob();` in 

```
static void foo() {
  bob();
}
```

The method returns an array of all matches mapped from the bound ID. Currently, the root node is always mapped to "root"

Todo:
 - [x] Write Tests
 - [x] Implement a client for vscode (@torshepherd)
 
 - [ ] Write documentation
 - [ ] Evaluate extendability for future feature requests.
 Possible Tasks for future PR's
 - allow search in all files
 - allow search in a list of files
 - allow blacklist xor whitelist based search
 - modifiable traversal kind
 - automatically bind root
 - autocompletion of nodes
 - generate matcher query from ast range

Febbe committed:
Add AST Search Capabilities to clangd
 - utilizes clang_query to find nodes in the AST
 - searches and returns the bound node names
 - added astSearchProvider to capabilities
 - added tests

HighCommander4 commented on clang-tools-extra/clangd/ClangdLSPServer.cpp:
```diff
@@ -76,6 +76,7 @@ std::optional<int64_t> decodeVersion(llvm::StringRef Encoded) {
 const llvm::StringLiteral ApplyFixCommand = "clangd.applyFix";
 const llvm::StringLiteral ApplyTweakCommand = "clangd.applyTweak";
 const llvm::StringLiteral ApplyRenameCommand = "clangd.applyRename";
+constexpr llvm::StringLiteral SearchASTMethod = "textDocument/searchAST";
```
nit: since this is only used once, and for consistency with other requests, just inline this string at the use site

HighCommander4 commented on clang-tools-extra/clangd/Protocol.h:
```diff
@@ -1451,6 +1452,19 @@ struct RenameParams {
 bool fromJSON(const llvm::json::Value &, RenameParams &, llvm::json::Path);
 llvm::json::Value toJSON(const RenameParams &);
 
+struct SearchASTArgs {
```
Please call this `SearchASTParams` for consistency with other names in this file

HighCommander4 commented on clang-tools-extra/clangd/ClangdServer.h:
```diff
@@ -260,6 +263,8 @@ class ClangdServer {
   void locateSymbolAt(PathRef File, Position Pos,
                       Callback<std::vector<LocatedSymbol>> CB);
 
+  void findAST(const SearchASTArgs &Args, Callback<BoundASTNodes> CB);
```
The convention at this layer seems to be to pass the input parameters separately, so something like:

```
void findAST(PathRef File, llvm::StringRef SearchQuery, Callback<BoundASTNodes> CB);
```

HighCommander4 commented on clang-tools-extra/clangd/ClangdServer.cpp:
```diff
@@ -810,6 +810,42 @@ void ClangdServer::locateSymbolAt(PathRef File, Position Pos,
+    CB(std::move(Result));
+  };
+
+  WorkScheduler->runWithAST("Definitions", Args.textDocument.uri.file(),
```
The first parameter is a string describing the task, so `"Definitions"` --> `"SearchAST"`

HighCommander4 commented:
@Febbe I've started looking at this, but I'd find it easier to reason about the code if I could trigger the new operation and play around with its behaviour.

Do you have an implementation in any client that you've tested this with?

Febbe committed:
Add capabilities to complete a matcher expression

 - adds a clang/completeASTMatcher method to the
   clangd LSP server, which returns a list of all
   available AST matcher expressions and the text
   to complete the current expression.
 - adds a test for the new method, which checks that
   the returned completions return the completion.
 - uses already existing code from ast-matchers.
 - mirrors behaviour of clang-query

Febbe committed:
fixes for first commit

Febbe committed:
fix for second

Febbe committed:
missing but used map include

Febbe committed:
size_t is not defined for json parsers

Febbe committed:
Merge branch 'main' into clangd_ast_search
